src/deepsparse/transformers/engines/nl_decoder_engine.py

```diff
@@ -190,23 +182,28 @@ def __call__(
         :param inp: The input to run the engine with. We expect a
             list of numpy arrays that contain the input ids,
             attention mask, and position ids (optionally)
+        :param kv_cache: The DecoderKVCache object that contains
+            the kv cache state
         :param val_inp: Whether the input is for validation or not
+
         :return: The generated token and corresponding logits
         """
         timer = self.timer_manager.current
-        if self.kv_cache:
+        if self.kv_cache_enabled:
             # if model has kv cache enabled, we need
             # to add the kv cache state to the input
-            inp = self.add_kv_cache_to_input(inp)
+            inp = self.add_kv_cache_to_input(inp, kv_cache)
 
         with timer.time(f"EXECUTE_ENGINE_SEQ_LEN_{self.sequence_length}"):
-            out = self.run(inp, val_inp)
+            out = self.run(inp, val_inp, kv_cache)
 
-        if self.kv_cache:
+        if self.kv_cache_enabled:
             with timer.time(TextGenerationTimings.KV_CACHE_UPDATE):
                 logits, *kv_cache_state = out
                 self.update_kv_cache(
-                    kv_cache_state=kv_cache_state, input_ids_len=self.input_ids_length
+                    kv_cache_state=kv_cache_state,
+                    input_ids_len=self.input_ids_length,
+                    kv_cache=kv_cache,
                 )
```

Comment:
I think this update_kv_cache function would make more sense as part of the DecoderKVCache? Since this class doesn't own the cache anymore it feels awkward to make changes to it here

Reply:
Yeah, I see what you mean here. On the other hand, all what `update_kv_cache` method does is convert a list of outputs into a dictionary (for that we need to have information about the names of the outputs that NLDecoderEngine holds) and then call kv_cache.update(...). I think that the responsibilities are fairly assigned here to every object.

Reply:
+1 to Sara's comment, we could move this function to DecoderKVCache and pass in relevant info, but this is not a blocker from my side

Reply:
gotcha, yeah not a blocker for me either